Python/import.c

```
@@ -1117,75 +1113,118 @@ list_frozen_module_names()
+typedef enum {
+    FROZEN_OKAY,
+    FROZEN_BAD_NAME,    // The given module name wasn't valid.
+    FROZEN_NOT_FOUND,   // It wasn't in PyImport_FrozenModules.
+    FROZEN_DISABLED,    // -X frozen_modules=off (and not essential)
+    FROZEN_EXCLUDED,    // The PyImport_FrozenModules entry has NULL "code".
+    FROZEN_INVALID,     // The PyImport_FrozenModules entry is bogus.
+} frozen_status;
+
+static inline void
+set_frozen_error(frozen_status status, PyObject *modname)
+{
+    switch (status) {
+        case FROZEN_BAD_NAME:
+        case FROZEN_NOT_FOUND:
+        case FROZEN_DISABLED:
+            PyErr_Format(PyExc_ImportError,
+                         "No such frozen object named %R",
+                         modname);
+            break;
+        case FROZEN_EXCLUDED:
+            PyErr_Format(PyExc_ImportError,
+                         "Excluded frozen object named %R",
+                         modname);
+            break;
+        case FROZEN_INVALID:
+            PyErr_Format(PyExc_ImportError,
+                         "Frozen object named %R is invalid",
+                         modname);
+            break;
+        case FROZEN_OKAY:
+            // There was no error.
+            break;
+        default:
+            Py_UNREACHABLE();
+    }
+}
+
+struct frozen_info {
+    PyObject *nameobj;
+    const char *data;
+    Py_ssize_t size;
+    bool is_package;
+};
+
+static frozen_status
+find_frozen(PyObject *nameobj, struct frozen_info *info)
 {
-    if (modname == NULL) {
-        return NULL;
+    if (info != NULL) {
+        info->nameobj = NULL;
+        info->data = NULL;
+        info->size = 0;
+        info->is_package = false;
+    }
+
+    if (nameobj == NULL || nameobj == Py_None) {
+        return FROZEN_BAD_NAME;
     }
-    const char *name = PyUnicode_AsUTF8(modname);
+    const char *name = PyUnicode_AsUTF8(nameobj);
     if (name == NULL) {
         PyErr_Clear();
-        return NULL;
+        return FROZEN_BAD_NAME;
```

Does `PyUnicode_AsUTF8()` not set an exception? That seems like something that should propagated up instead of silenced.

---

It does set an exception, hence `PyErr_Clear()`.  I added that in gh-28320, where `PyUnicode_AsUTF8()` replaced the existing usage of `_PyUnicode_EqualToASCIIString()`.  The latter either asserts in error cases or ignores the errors, so I followed suit, effectively matching the existing behavior of the code.  I don't mind propagating the error but didn't want to change behavior.

---

I've added a note.